# 💡 아이템 75 예외의 상세 메시지에 실패 관련 정보를 담으라

## 개요
* 예외를 잡지 못해 프로그램이 실패하게 된다면 **자바는 예외의 스택 추적 정보를 자동으로 출력**한다.
* 스택 추적은 예외 객체의 toString 메소드를 호출해 얻는 문자열로, 보통은 예외의 클래스 이름 뒤에 상세 메세지가 붙는 형태입니다. **(오류를 분석하는 유일한 정보)**
* 정확한 오류 파악을 위해 예외의 실패 원인에 관한 정보를 가능한 많이 담아 반환하는 일은 중요합니다. 



##   
### ✅ 예시코드
```java
public class Calculation {
    private int dividend;
    private int divisor;

    public double divide() {
        try {
            // 나눗셈 연산 수행
            return dividend / divisor;
        } catch (ArithmeticException e) {
            // 실패 순간에 관련된 매개변수와 필드 값 포함
            String message = String.format("나눗셈 오류 발생: dividend=%d, divisor=%d", dividend, divisor);
            throw new ArithmeticException(message);
        }
    }
}
```
* 위 예시코드와 같이 실패 순간을 좀 더 정확히 포착하려면 발생한 **예외에 관여된 모든 매개변수와 필드의 값을 실패 메세지에 담아야** 합니다.
* 이때 예외의 **상세 메세지**와 **최종 사용자에게 보여줄 오류 메세지**를 혼동해서는 안됩니다.
* 최종 사용자에게는 친절한 안내문구를 보여줘야하는 반면, 예외 메세지는 가독성보다는 담긴 내용이 중요합니다.

##   
### ✅ 예외 접근자 메소드
```java
public class DataException extends Exception {
    private int errorCode;
    private String errorMessage;

    public DataException(int errorCode, String errorMessage) {
        this.errorCode = errorCode;
        this.errorMessage = errorMessage;
    }

    public int getErrorCode() {
        return errorCode;
    }

    public String getErrorMessage() {
        return errorMessage;
    }
}
```
* 위 예시 코드는 실패와 관련된 정보를 얻을 수 있는 **접근자 메소드를 제공**하고 있습니다.
* DataException은 errorCode와 errorMessage를 가지는 사용자 정의 예외 클래스입니다.
* 예외 객체를 생성할 떄 실패와 관련된 정보를 전달하고 접근자 메소드를 통해 정보를 얻을 수 있습니다.


### main 메소드
```java
   public static void main(String [] args) throws DataException{
        try {
            // 데이터 처리 로직
            // ...

            // 실패 발생 시 예외를 던집니다.
            throw new DataException(500, "데이터 처리 실패");
        } catch (DataException e) {
            // 실패에 대한 정보 활용
            System.out.println("에러 코드: " + e.getErrorCode());
            System.out.println("에러 메시지: " + e.getErrorMessage());
        }
        
    }
```
* DataException을 던질 때 생성자에 실패 정보들을 저장하고, 저장된 객체를 활용하여 출력하는 예시입니다.


##
### 정리
* **예외**는 실패와 관련한 정보를 얻을 수 있는 **접근자 메소드를 적절히 제공하는 것**이 좋습니다.
* **포착한 실패 정보는 예외 상황을 복구**하는데 유용할 수 있습니다.
* 당연히 기본적인 예외 상황에서의 toString이 반환된 정보가 정확할 수 있지만, **접근자 메소드를 통해 예외 상세 정보를 알려주어 좀더 쉬운 오류 해결**을 이룰 수 있습니다.
